log.c

```diff
@@ -0,0 +1,55 @@
+#define _GNU_SOURCE
+#include "log.h"
+#include "common.h"
+
+#include <pthread.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+__attribute__ ((format (printf, 5, 6))) status_t
+message_impl (const char *file_name, const char *func_name, int line,
+              log_level_t level, const char *msg, ...)
+{
+  static log_level_t log_level = LL_OFF;
+#ifdef LOG_LEVEL
+#define LOG_LEVEL_MACRO(str) LL_##str
+#define EXPAND_LOG_LEVEL_MACRO(str) LOG_LEVEL_MACRO (str)
+  log_level = EXPAND_LOG_LEVEL_MACRO (LOG_LEVEL);
+#endif
+
+  if (msg == NULL)
+    {
+      log_level = level;
+      return (S_SUCCESS);
+    }
+
+  if (level < log_level)
+    return (S_SUCCESS);
+
+  char *log;
+  if (asprintf (&log, "(%s %s %d) ", file_name, func_name, line) < 0)
```

Comment:
Let's use statically allocated buffers of some reasonable size. Something like 
char log[1 << 6];
Both for the log[] and for the message[].
Dynamics allocation here is an unnecessary overhead.

Reply:
Done